File: 0140-word-break-ii/0140-word-break-ii.java
class Solution {
    public List<String> wordBreak(String sentence, List<String> wordDict) {
        int n = sentence.length();
        List<List<String>>[] dp = new List[n + 1];
        dp[0] = new ArrayList<>();
        dp[0].add(new ArrayList<>());

        for (int i = 0; i <= n; i++) {
            if (dp[i] != null) {
                for (String word : wordDict) {
                    int m = word.length();
                    if (i + m <= n && sentence.substring(i, i + m).equals(word)) {
                        for (List<String> combination : dp[i]) {
                            List<String> newCombination = new ArrayList<>(combination);
                            newCombination.add(word);
                            if (dp[i + m] == null) {
                                dp[i + m] = new ArrayList<>();
                            }
                            dp[i + m].add(newCombination);
                        }
                    }
                }
            }
        }

        List<String> result = new ArrayList<>();
        if (dp[n] != null) {
            for (List<String> wordGroup : dp[n]) {
                result.add(String.join(" ", wordGroup));
            }
        }

        return result;
    }
}

File: 0140-word-break-ii/0140-word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:

        def backtrack(target, curr):
            if target == '':
                result.append(curr[:])
                return
            
            for word in wordDict:
                if target.startswith(word):
                    curr.append(word)
                    backtrack(target[len(word) : ], curr)
                    curr.pop()

        result = []
        backtrack(s, [])
        
        sentences = []
        for word_list in result:
            sentence = " ".join(word_list)
            sentences.append(sentence)
        
        return sentences
        